plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageClient.java

```diff
@@ -0,0 +1,412 @@
+
+import com.google.api.client.googleapis.batch.BatchRequest;
+import com.google.api.client.googleapis.batch.json.JsonBatchCallback;
+import com.google.api.client.googleapis.json.GoogleJsonError;
+import com.google.api.client.googleapis.json.GoogleJsonResponseException;
+import com.google.api.client.http.HttpHeaders;
+import com.google.api.client.http.InputStreamContent;
+import com.google.api.services.storage.Storage;
+import com.google.api.services.storage.model.Bucket;
+import com.google.api.services.storage.model.Objects;
+import com.google.api.services.storage.model.StorageObject;
+import org.elasticsearch.SpecialPermission;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.blobstore.BlobMetaData;
+import org.elasticsearch.common.blobstore.support.PlainBlobMetaData;
+import org.elasticsearch.common.component.AbstractComponent;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.concurrent.CountDown;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.security.AccessController;
+import java.security.PrivilegedActionException;
+import java.security.PrivilegedExceptionAction;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Spliterator;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import static java.net.HttpURLConnection.HTTP_FORBIDDEN;
+import static java.net.HttpURLConnection.HTTP_NOT_FOUND;
+
+/**
+ * Client for Google Cloud Storage service.
+ * <p>
+ * Some useful errors description can be found at
+ * https://cloud.google.com/storage/docs/reference-status
+ */
+public class GoogleCloudStorageClient extends AbstractComponent {
+
+    /**
+     * Google Cloud Storage batch requests are limited to 1000 operations
+     **/
+    private static final int MAX_BATCHING_REQUESTS = 999;
+
+    private final Storage client;
+
+    public GoogleCloudStorageClient(Settings settings, Storage client) {
+        super(settings);
+        this.client = client;
+    }
+
+    /**
+     * Return true if the given bucket exists
+     *
+     * @param bucketName name of the bucket
+     * @return true if the bucket exists, false otherwise
+     */
+    public boolean doesBucketExist(String bucketName) throws IOException {
+        return doPrivileged(() -> {
+            try {
+                Bucket bucket = client.buckets().get(bucketName).execute();
+                if (bucket != null) {
+                    return Strings.hasText(bucket.getId());
+                }
+            } catch (GoogleJsonResponseException e) {
+                if (e.getStatusCode() == HTTP_FORBIDDEN) {
+                    throw new IOException("Unable to get bucket [" + bucketName + "] because of insufficient privileges");
```

Comment:
I am wondering if it's safe to pretend that bucket doesn't exist when we don't have sufficient privileges. I think it might be safer to just bailout here.